File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int get_length(long a)
{
    int length = 0;
    while (a > 0)
    {
        a /= 10;
        length ++;
    }
    return length;
}

int main(void)
{
    long number = get_long("Number: ");
    int check;
    long start = number;
    int sum1 = 0;
    int sum2 = 0;
    int mod1;
    int mod2;
    int digit1; // For number that has 2 digits after times 2
    int digit2;

    int length = get_length(number);

    // Length validation
    if (length != 13 && length != 15 && length != 16)
    {
        printf("INVALID\n");
        return 0;
    }

    // Extract each digits of the credit card numbers
    do
    {
        mod1 = number % 10;
        number = number / 10;
        sum1 += mod1;

        mod2 = (number % 10) * 2;
        number = number / 10;
        digit1 = mod2 % 10;
        digit2 = mod2 / 10;
        sum2 = sum2 + digit1 + digit2;
    }
    while (number > 0);

    check = sum1 + sum2;

    // Validate check sum
    if (check % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    // Get first 2 digits and validate card type
    do
    {
        start /= 10;
    }
    while (start > 100);

    if (start / 10 == 4)
    {
        printf("VISA\n");
    }
    else if ((start / 10 == 5) && (0 < start % 10 && start % 10 < 6))
    {
        printf("MASTERCARD\n");
    }
    else if ((start / 10 == 3) && (start % 10 == 4 || start % 10 == 7))
    {
        printf("AMEX\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: pset6/credit/credit.py
from cs50 import get_int, get_string


def main():
    number = get_string("Number: ")
    length = len(number)
    number = int(number)
    sum1 = 0
    sum2 = 0
    first = number

    if length != 13 and length != 15 and length != 16:
        print("INVALID")
        return 0

    while number > 0:
        mod1 = number % 10
        number = number // 10
        sum1 += mod1

        mod2 = (number % 10) * 2
        number = number // 10
        digit1 = mod2 % 10
        digit2 = mod2 // 10
        sum2 = sum2 + digit1 + digit2

    check = sum1 + sum2

    if check % 10 != 0:
        print("INVALID")
        return 0

    while first > 100:
        first = first // 10

    if first // 10 == 4:
        print("VISA")
    elif first // 10 == 5 and (0 < first % 10 and first % 10 < 6):
        print("MASTERCARD")
    elif first // 10 == 3 and (first % 10 == 4 or first % 10 == 7):
        print("AMEX")
    else:
        print("INVALID")


main()